replset-consistency/dbcheck.js

```
@@ -154,34 +154,93 @@ function getDBCheckCount(readPref) {
+    { $match: { operation: "dbCheckStart", timestamp: { $gte: startDate } } },
+    { $count: "dbCheckStartCount" },
+  ]);
+
+  if (curr.hasNext()) {
+    let my_count = curr.next().dbCheckStartCount;
+    node.dbCheckStartCount = my_count;
+    return my_count;
+  }
+  return 0;
+}
+
+// Authenticates into every data-bearing secondary nodes to query healthlog 
+// dbCheckStart entries
+//
+// If authInfo is not provided, behavior defaults to old healthlog rollover check 
+// that is non-deterministic. We don't know if we'll select the same secondary 
+// over and over again.
 //
 function checkRollOver() {
+  let config = rs.config();
+  let nodelist = [];
+
+  
   try {
     primaryCount = getDBCheckCount("primary")
   } catch (error) {
     printFunction(error);
   }
-  for (let i = 0; i < ((getWriteConcern() - 1) * 2); i++) {
-    try {
-      let tcount = getDBCheckCount("secondary");
-      if (i == 0) {
-        secondaryCount = tcount;
-      } else {
-        secondaryCount = Math.min(secondaryCount, tcount);
+
+  if (typeof authInfo !== "undefined") {
+    uriOptions = authInfo.uriOptions;
+    delete authInfo.uriOptions;
+  
+    for (let member of config.members) {
+        let conn = new Mongo("mongodb://" + member.host + "/?" + uriOptions);
+        conn.setSecondaryOk(true);
+        if (member.arbiterOnly) {
+            conn.close();
+        } else {
+            if (authInfo) {
+              conn.auth(authInfo);
+            }
+            nodelist.push({_id: member._id, connection: conn, host: member.host});
+        }
+    }
+    
+    for (let i = 0; i < nodelist.length; i++) {
+      let nodeInfo = nodelist[i];
+      try {   
+        let tcount = getDBCheckCountByNode(nodeInfo)
+        if (i == 0) {
+          secondaryCount = tcount;
+        } else {
+          secondaryCount = Math.min(secondaryCount, tcount);
+        }
+      } catch (error) {
+        printFunction(error);
```

Let's add a more explicit error here that we have failed to check which node.

I also wonder if we should fail harder if, say, a majority of nodes cannot be reached?